led_strip/src/led_strip_spi_dev.c

```diff
@@ -0,0 +1,212 @@
+    ESP_GOTO_ON_FALSE(led_config && spi_config && ret_strip, ESP_ERR_INVALID_ARG, err, TAG, "invalid argument");
+    ESP_GOTO_ON_FALSE(led_config->led_pixel_format < LED_PIXEL_FORMAT_INVALID, ESP_ERR_INVALID_ARG, err, TAG, "invalid led_pixel_format");
+    uint8_t bytes_per_pixel = 3;
+    if (led_config->led_pixel_format == LED_PIXEL_FORMAT_GRBW) {
+        bytes_per_pixel = 4;
+    } else if (led_config->led_pixel_format == LED_PIXEL_FORMAT_GRB) {
+        bytes_per_pixel = 3;
+    } else {
+        assert(false);
+    }
+    uint32_t mem_caps = MALLOC_CAP_DEFAULT;
+    if (spi_config->flags.with_dma) {
+        // DMA descriptors must be placed in internal SRAM
+        mem_caps |= MALLOC_CAP_32BIT | MALLOC_CAP_DMA;
+    }
+    spi_strip = heap_caps_calloc(1, sizeof(led_strip_spi_obj) + led_config->max_leds * bytes_per_pixel * LED_COLOR_BUF, mem_caps);
+
+    //spi_strip = calloc(1, sizeof(led_strip_spi_obj) + led_config->max_leds * bytes_per_pixel * LED_COLOR_BUF);
+    ESP_GOTO_ON_FALSE(spi_strip, ESP_ERR_NO_MEM, err, TAG, "no mem for spi strip");
+    uint32_t resolution = spi_config->resolution_hz ? spi_config->resolution_hz : LED_STRIP_SPI_DEFAULT_RESOLUTION;
+
+    spi_strip->spi_host = spi_config->host_channel;
+    // for backward compatibility, if the user does not set the clk_src, use the default value
+    spi_clock_source_t clk_src = SPI_CLK_SRC_DEFAULT;
+    if (spi_config->clk_src) {
+        clk_src = spi_config->clk_src;
+    }
+
+    spi_bus_config_t spi_bus_cfg = {
+        .mosi_io_num = led_config->strip_gpio_num,
+        .miso_io_num = -1,
+        .sclk_io_num = -1,
+        .quadwp_io_num = -1,
+        .quadhd_io_num = -1,
+        .max_transfer_sz = led_config->max_leds * bytes_per_pixel * LED_COLOR_BUF,
+    };
+    ESP_GOTO_ON_ERROR(spi_bus_initialize(spi_strip->spi_host, &spi_bus_cfg, spi_config->flags.with_dma ? SPI_DMA_CH_AUTO : SPI_DMA_DISABLED), err, TAG, "create SPI bus failed");
+
+#if !CONFIG_IDF_TARGET_ESP32
+    //SPI_D_POL : This bit is used to set the idle polarity of MOSI. 1：high；0：low (ESP32 does not have this register, and its default polarity is low)
+    spi_dev_t *hw;
+    hw = SPI_LL_GET_HW(spi_strip->spi_host);
+    hw->ctrl.d_pol = 0;
+#endif
+
+
+    if (led_config->flags.invert_out == true) {
+        //Inverse the polarity of MOSI(not recommand)
+        esp_rom_gpio_connect_out_signal(led_config->strip_gpio_num, spi_periph_signal[spi_strip->spi_host].spid_out, true, false);
+    }
+
+    spi_device_interface_config_t spi_dev_cfg = {
+        .clock_source = clk_src,
+        .command_bits = 0,
+        .address_bits = 0,
+        .dummy_bits = 0,
+        .clock_speed_hz = resolution,
+        .duty_cycle_pos = 128,
+        .mode = 0,
+        .spics_io_num = -1,
+        .queue_size = LED_STRIP_SPI_DEFAULT_TRANS_QUEUE_SIZE,
+    };
+
+    ESP_RETURN_ON_ERROR(spi_bus_add_device(spi_strip->spi_host, &spi_dev_cfg, &spi_strip->spi_device), TAG, "Failed to add spi device");
```

Comment:
if return here, then `spi_strip` which was allocated before will be leaked forever.